nupic/research/frameworks/pytorch/datasets/permuted_mnist.py

```diff
@@ -0,0 +1,146 @@
+    """
+
+    # Permutations should be the same whether this class is instantiated for training
+    # or testing
+    permutations = None
+
+    def __init__(self, num_tasks, train, root=".", target_transform=None,
+                 download=False):
+
+        data_transform = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize((0.13062755,), (0.30810780,)),
+        ])
+        super().__init__(root=root, train=train, transform=data_transform,
+                         target_transform=target_transform, download=download)
+
+        self.num_tasks = num_tasks
+
+        # Generate num_tasks random permutations - the first one is the identity
+        # permutation (i.e., regular MNIST), represented below as `None`
+        if PermutedMNIST.permutations is None:
+            PermutedMNIST.permutations = [
+                torch.randperm(784) for task_id in range(1, num_tasks)
+            ]
+            PermutedMNIST.permutations.insert(0, None)
+
+        self.permutations = PermutedMNIST.permutations
+
+    def __getitem__(self, index):
+        """
+        Returns an (image, target) pair.
+
+        In particular, this method retrieves an MNIST image, and based on the value of
+        `index`, determines which pixel-wise permutation to apply. Target values are
+        also scaled to be unique to each permutation.
+        """
+        img, target = super().__getitem__(index % len(self.data))
+
+        # Determine which task `index` corresponds to
+        task_id = self.get_task_id(index)
+
+        # Apply permutation to `img`
+        img = permute(img, self.permutations[task_id])
+
+        # Since target values are not shared between tasks, `target` should be in the
+        # range [0 + 10 * task_id, 9 + 10 * task_id]
+        target += 10 * task_id
+        return img, target
+
+    def __len__(self):
+        return self.num_tasks * len(self.data)
+
+    @property
+    def processed_folder(self):
+        return os.path.join(self.root, "MNIST", "processed")
+
+    def get_task_id(self, index):
+        return index // len(self.data)
+
+
+class ContextDependentPermutedMNIST(PermutedMNIST):
+    """
+    A variant of permutedMNIST where each permutation (i.e., 'task') is associated with
+    a context: a binary sparse vector. The `__getitem__` method returns the context
+    vector along with the data sample and target.
+    """
+
+    # Context vectors should be the same whether this class is instantiated for
+    # training or testing
+    contexts = None
+
+    def __init__(self, num_tasks, dim_context, train, root=".", target_transform=None,
+                 download=False):
+
+        super().__init__(num_tasks, train, root, target_transform, download)
+
+        # Generate context vectors if `contexts` isn't already defined
+        if ContextDependentPermutedMNIST.contexts is None:
+            ContextDependentPermutedMNIST.contexts = generate_context_vectors(
```

Comment:
Same coment applues here, why is `contexts` a class variable

Reply:
Same reason why `permutations` is a class variable.